network/kademlia.go

```diff
@@ -109,11 +112,92 @@ func NewKademlia(addr []byte, params *KadParams) *Kademlia {
 	if params == nil {
 		params = NewKadParams()
 	}
-	return &Kademlia{
-		base:      addr,
-		KadParams: params,
-		addrs:     pot.NewPot(nil, 0),
-		conns:     pot.NewPot(nil, 0),
+	k := &Kademlia{
+		base:            addr,
+		KadParams:       params,
+		capabilityIndex: make(map[string]*capabilityIndex),
+		addrs:           pot.NewPot(nil, 0),
+		conns:           pot.NewPot(nil, 0),
+	}
+	k.RegisterCapabilityIndex("full", *fullCapability)
+	k.RegisterCapabilityIndex("light", *lightCapability)
+	return k
+}
+
+// RegisterCapabilityIndex adds an entry to the capability index of the kademlia
+// The capability index is associated with the supplied string s
+// Any peers matching any bits set in the capability in the index, will be added to the index (or removed on removal)
+func (k *Kademlia) RegisterCapabilityIndex(s string, c capability.Capability) error {
+	if s == "" {
+		return errors.New("Cannot add index with empty string key")
+	} else if _, ok := k.capabilityIndex[s]; ok {
+		return fmt.Errorf("Capability index '%s' already exists", s)
+	}
+	log.Debug("Registered cap index", "s", s, "c", c)
+	k.capabilityIndex[s] = NewCapabilityIndex(c)
+	return nil
+}
+
+// adds a peer to any capability indices it matches
+func (k *Kademlia) addToCapabilityIndex(p interface{}) {
+	var ok bool
+	var eAddr *BzzAddr
+	var ePeer *Peer
+	ePeer, ok = p.(*Peer)
+	if ok {
+		eAddr = ePeer.BzzAddr
+	} else {
+		eAddr = p.(*entry).BzzAddr
+	}
+	for s, idxItem := range k.capabilityIndex {
+		for _, vCap := range eAddr.Capabilities.Caps {
+			if idxItem.Id != vCap.Id {
+				continue
+			}
+			if vCap.Match(idxItem.Capability) {
+				log.Trace("Added peer to capability index", "conn", ok, "s", s, "v", vCap, "p", p)
+				if ok {
+					k.capabilityIndex[s].conns, _, _ = pot.Add(idxItem.conns, ePeer, Pof)
+				} else {
+					k.capabilityIndex[s].addrs, _, _ = pot.Add(idxItem.addrs, newEntry(eAddr), Pof)
+				}
+			}
+		}
+	}
+}
+
+// removes a peer from any capability indices it matches
+func (k *Kademlia) removeFromCapabilityIndex(p interface{}, disconnectOnly bool) {
+	var ok bool
+	var eAddr *BzzAddr
+	var ePeer *Peer
+	ePeer, ok = p.(*Peer)
+	if ok {
+		eAddr = ePeer.BzzAddr
+	} else if disconnectOnly {
+		return
+	} else {
+		eAddr = p.(*entry).BzzAddr
+	}
+	for s, idxItem := range k.capabilityIndex {
+		if ok {
```

Comment:
The code in these two ifs seems almost the same. Could not be it improved to be more compact?
Maybe a new function idxItem.swap(p, Pof)?

```go
func (ci *capabilityIndex) swap(p interface{}, s string) {
	var index *pot.Pot
	switch p.(type) {
	case *Peer:
		index = ci.conns
	case *entry:
		index = ci.addrs
	}
	newIndx, _, found, _ := pot.Swap(index, p, Pof, func(_ pot.Val) pot.Val {
		return nil
	})
	if found {
		log.Trace("Removed peer from capability index", "s", s, "p", p)
		switch p.(type) {
		case *Peer:
			ci.conns = newIndx
		case *entry:
			ci.addrs = newIndx
		}
	}
}
```

Reply:
We could, but that's indirection per index instead of just once before the loop. This code does not need to be the most performant, perhaps, but I'm also not sure how much more clarity the change brings.

Reply:
In fact that swap method (although I should call it remove) could be split in two methods removeAddress and removeConnection, so no need of switch at the start of the method is needed: 
```go
func (ci *capabilityIndex) removeAddress(addr *BzzAddr, s string) {
	newIndx, _, found, _ := pot.Swap(ci.addrs, addr, Pof, func(_ pot.Val) pot.Val {
		return nil
	})
	if found {
		ci.addrs = newIndx
	}
}

func (ci *capabilityIndex) removeConnection(conn *Peer, s string) {
	newIndx, _, found, _ := pot.Swap(ci.conns, conn, Pof, func(_ pot.Val) pot.Val {
		return nil
	})
	if found {
		ci.conns = newIndx
	}
}


```

Reply:
That is true. But I'm still not convinced it contributes to increased clarity, as now you have the two function calls separately anyway :) I would prefer to leave it as it is, if you don't mind.

Reply:
Ok, I think the problem is more like you said that we are using two different types in the indexes instead of always using *entry.